### PR TITLE
CEDS-2278 - only show 'OSR' pages if procedure code is '1042'

### PR DESCRIPTION
**Pull request by SteveSugden**

### CODE REVIEW

File: app/controllers/declaration/FiscalInformationController.scala [hunk 1]
```diff
@@ -43,10 +44,18 @@ class FiscalInformationController @Inject()(
...
+      request.cacheModel
+        .itemBy(itemId)
+        .flatMap(_.procedureCodes)
+        .flatMap(_.procedureCode)
+        .exists(code => !ProcedureCodesData.osrProcedureCodes.contains(code))
 
     if (fastForward && cacheContainsFiscalReferenceData) {
       navigator.continueTo(routes.AdditionalFiscalReferencesController.displayPage(mode, itemId))
+    } else if (fastForward && cacheItemIneligibleForOSR) {
+      navigator.continueTo(routes.ProcedureCodesController.displayPage(mode, itemId))
```
Comment (PatrykRudnicki) [inline app/controllers/declaration/FiscalInformationController.scala:58]: Just to confirm this redirection is correct? I thought that ProcedureCodes is a page before Fiscal Information

Comment (SteveSugden) [reply app/controllers/declaration/FiscalInformationController.scala:58]: That is for "back link" navigation.   Perhaps the "fastforward" variable could be better named "back navigating"?

The logic is when coming back from the next page (Commodity Code) we direct to FiscalInfo.   We then either redirect one page forward to the VAT page if there is existing fiscal data (old implementation), re-direct to the page before if OSR not valid due to existing Procedure Code (new logic) or stay here.

I've tested this manually and it all appears to work.   It is a bit "clunky" though.

Comment (SteveSugden) [reply app/controllers/declaration/FiscalInformationController.scala:58]: There are no existing tests in the FiscalControllerSpec for "fast-forward" BTW - adding those now including the new ProcedureCode invalid logic.

Comment (PatrykRudnicki) [reply app/controllers/declaration/FiscalInformationController.scala:58]: Ahh, I forgot about it. Thanks for explanation!

That's right.

File: app/models/declaration/ExportItem.scala [hunk 2]
```diff
@@ -40,19 +40,25 @@ case class ExportItem(
...
-    if (fiscalInformation.exists(_.onwardSupplyRelief == yes)) additionalFiscalReferencesData.isDefined
-    else fiscalInformation.isDefined
+  private def isProcedureCodesAndFiscalInformationComplete = {
+
+    def isFiscalInformationCompleted: Boolean =
+      if (hasFiscalReferences) additionalFiscalReferencesData.isDefined
+      else fiscalInformation.isDefined
+
+    procedureCodes.flatMap(_.procedureCode).isDefined &&
+    (procedureCodes.flatMap(_.procedureCode).exists(code => !ProcedureCodesData.osrProcedureCodes.contains(code)) || isFiscalInformationCompleted)
```
Comment (PatrykRudnicki) [inline app/models/declaration/ExportItem.scala:60]: I'm just wondering if the `isFiscalInformationCompleted` must be in exists method or outside it.

Comment (SteveSugden) [reply app/models/declaration/ExportItem.scala:60]: I this this is right - i.e. outside.   The procedure code be defined and either not be "1042" or fiscal information must be provided.

**Pull request merged**